File: F# Workspace/Asignacion 5 (Comparacion de Lenguajes)/Asignacion 5 (Comparacion de Lenguajes)/employee_payroll.fs
(*
 * employee_payroll.fs
 * By Saulo Almeda Nieves
 * 841-13-0170
 * 11/27/2015
 * This program creates a list of at least five different kinds of employees
 * and displays their data, including their weekly salary with the appropiate
 * format.
 *)

//Defines a discriminated union type for Employee.
type Employee =
    | HourlyEmployee of string * string * float * float
    | SalesEmployee of string * string * float * float

//Calculates the weekly salary of the given employee.
let weekly = function
    | HourlyEmployee (_, _, payrate, hrsworked) ->
        if hrsworked > 40.00 then ((hrsworked - 40.00) * payrate * 1.5) + (40.00 * payrate)
        else payrate * hrsworked
    | SalesEmployee (_, _, commrate, salesamount) ->
        commrate * salesamount

// show_emplyee : Employee -> unit
// Displays the data of the given employee.
let show_employee emp =
    match emp with
    | HourlyEmployee (id, name, payrate, hrsworked) ->
        printfn "ID: %s" id
        printfn "Name: %s" name
        printfn "Pay rate: %.2f" payrate
        printfn "Hours worked: %.2f" hrsworked
    | SalesEmployee (id, name, commrate, salesamount) ->
        printfn "ID: %s" id
        printfn "Name: %s" name
        printfn "Commission rate: %.2f" (commrate * 100.00)
        printfn "Sales amount: $%.2f" salesamount

    printfn "Weekly salary: $%.2f\n" (weekly emp)

//Serves as starting point of the app.
let start () =
    let lst = [
        HourlyEmployee ("1234", "Saulo", 7.25, 35.00); 
        SalesEmployee ("2345", "Saul", 0.50, 6000.00);
        HourlyEmployee ("3456", "Raquel", 8.00, 45.00);
        SalesEmployee ("4567", "Rebeca", 0.30, 1500.00);
        HourlyEmployee ("5678", "Judith", 32.00, 40.00)]

    printfn "These are the employees:"
    List.iter (fun elem -> printfn "Employee Data..."; show_employee elem) lst

File: C_Projects/Asignacion #3 (Comparacion de Lenguajes)/employee_payroll.c
/*  Saulo Almeda Nieves
 *  841-13-0170
 *  11 / 2 /2015
 *  employee_payroll.c
 *  This program creates an array of at least five different kinds of employees
 *  and displays their data, including their weekly salary with the appropriate
 *  format.
 */

//Includes

//Defines
#define ID_SIZE 4
#define NAME_SIZE 20

//Defines Data Structures
typedef enum {HOURLY, SALES} EmployeeKind;

typedef struct {
   float pay_rate,
         hours_worked;
} HourlyEmployee;

typedef struct {
   float commission_rate,
         sales_amount;
} SalesEmployee;

typedef struct {
   EmployeeKind kind;
   char employee_id[ID_SIZE + 1],
        full_name[NAME_SIZE + 1];
   union {
      HourlyEmployee hourly;
      SalesEmployee sales;
   };
} Employee;

//Function Prototypes
void display_data(const Employee *emp);
char *kind_to_str(EmployeeKind kind);

//Main
int main(void) {
    Employee arr[] = {
        {HOURLY, "1234", "Saulo Almeda Nieves", .hourly = {5.0, 10}},
        {HOURLY, "2468", "Raquel Almeda Nieves", .hourly = {7.0, 20}},
        {HOURLY, "4812", "Rebeca Almeda Nieves", .hourly = {8.0, 50}},
        {SALES, "8162", "Judith Nieves Rivera", .sales = {.20, 100}},
        {SALES, "1621", "Saul Omar Almeda Cruz", .sales = {.40, 200}}};

    int arr_length = sizeof(arr) / sizeof(Employee);

    printf("These are the employees:\n\n");

    for(int idx = 0; idx < arr_length; idx++) {
        printf("Employee #%d\n\n", idx + 1);
        display_data(&arr[idx]);
    }
    return 0;
}//end of main

//Dispplay all the data of the given Employee
void display_data(const Employee *emp) {
    printf("Kind: %s\nID: %s\nFull Name: %s\n", kind_to_str(emp->kind),
           emp->employee_id, emp->full_name);
    switch (emp->kind) {
        case HOURLY:
            if(emp->hourly.hours_worked > 40) {
                printf("Weekly Salary: $%.2f\n\n", (emp->hourly.pay_rate + 1.5) * emp->hourly.hours_worked);
                break;
            }
            printf("Weekly Salary: $%.2f\n\n", emp->hourly.pay_rate *
                   emp->hourly.hours_worked);
            break;
        case SALES:
            printf("Weekly Salary: $%.2f\n\n", emp->sales.commission_rate * emp->sales.sales_amount);
            break;
    }
}

//Returns the equivalent enum value in a string
char *kind_to_str(EmployeeKind kind) {
    char *str[] = {"Hourly", "Sales"};
    return str[kind];
}
